src/odbc_connection.cpp

```diff
@@ -2844,108 +2913,399 @@ fetch_and_store
+              data->sqlReturnCode =
+              SQLGetData
+              (
+                data->hSTMT,
+                column_index + 1,
+                data->columns[column_index]->bind_type,
+                data->long_data_buffer,
+                data->long_data_buffer_size,
+                &data->long_data_string_length_or_indicator
+              );
+              printf("SQLGetData result: %d\n", data->sqlReturnCode);
+              if (!SQL_SUCCEEDED(data->sqlReturnCode))
+              {
+                return data->sqlReturnCode;
+              } 
+
+              // If the data is null, simply indicate and continue to the next
+              // column
+              if (data->long_data_string_length_or_indicator == SQL_NULL_DATA)
+              {
+                row[column_index].size = SQL_NULL_DATA;
+                continue;
+              } else if (data->long_data_string_length_or_indicator == SQL_NO_TOTAL) 
+              {
+                SQLLEN data_to_copy_length = 0;
+                while (data->long_data_string_length_or_indicator == SQL_NO_TOTAL)
                 {
-                  row[column_index].size = data->bound_columns[column_index].length_or_indicator_array[row_index];
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size]();
-                  memcpy(
-                    row[column_index].char_data,
-                    (SQLCHAR *)data->bound_columns[column_index].buffer + row_index * data->columns[column_index]->buffer_size,
-                    row[column_index].size
+                  switch(data->columns[column_index]->bind_type)
+                  {
+                    case SQL_C_BINARY:
+                      data_to_copy_length = data->long_data_buffer_size;
+                      row[column_index].char_data =
+                      (SQLCHAR *)
+                      realloc
+                      (
+                        row[column_index].char_data,
+                        row[column_index].size + data_to_copy_length
+                      );
+                      memcpy
+                      (
+                        row[column_index].char_data + row[column_index].size,
+                        data->long_data_buffer,
+                        data_to_copy_length
+                      );
+                    case SQL_C_WCHAR:
+                      data_to_copy_length = strlen16((const char16_t *)data->long_data_buffer) * 2;
+                      row[column_index].wchar_data =
+                      (SQLWCHAR *)
+                      realloc
+                      (
+                        row[column_index].wchar_data,
+                        row[column_index].size + data_to_copy_length
+                      );
+                      memcpy
+                      (
+                        row[column_index].wchar_data + row[column_index].size,
+                        data->long_data_buffer,
+                        data_to_copy_length
```

Comment:
While this does work, it is a lot of potentially duplicate memcpy calls. Instead, it would be more efficient to retrieve the whole buffer in to the target data buffer instead of having long_data_buffer at all.

This gives a _rough_ approximation:
```c
size_t size = initial size;
char* buffer = malloc(size);
size_t offset = 0;
SQLLEN indicator = size;

while(1) {
    SQLGetData(type, &buffer[offset], indicator);

    // We read all the data
    if (indicator > 0 && indicator < (size - offset)) break;
    
    // Adjust the offset based on how much was returned
    // This code needs to be adjusted to work with binary and wide data
    offset += strlen(&buffer[offset);

    // Resize the buffer to fit more data
    if (indicator == SQL_NO_TOTAL || indicator > size) {
       size_t newsize;
       if (indicator == SQL_NO_TOTAL) newsize = size + reasonable extra;
       else newsize = indicator + 1;

       char* newptr = realloc(buffer, newsize);
       if(!newptr) handle error and return
       buffer = newptr;
       size = newsize;
    }
}
```

Any time the buffers are realloc'd, they likely need to copy all the existing data to the new pointer (unless there was space available in the current allocation) and the getdata buffer needs to be copied over, then it loops again to get more data where the process repeats. Additionally, the getdata buffer can be realloc'd, causing potentially further copies.

Instead, we can use the target buffer directly and realloc only it, except we treat it as two halves: the first half contains valid data while the second half is used as target for getdata (equivalent to long_data_buffer).